livehist.1
.TH LIVEHIST 1
.SH NAME
livehist \- A real-time histogram for data streams
.SH SYNOPSIS
.B livehist\fR [\fIoptions\fR]
.SH DESCRIPTION
.B livehist\fR displays a real-time histogram of the data stream piped into the tool.  
Symbols are separated by newlines, i.e., every line is seen as one symbol in the histogram.
After every read symbol, the histogram is updated.
The output can either be as relative or absolute number, or more graphical using percentage bars. 

With the \fI\-a\fR option, \fBlivehist\fR can also ensure that the histogram output is not interrupted when aborting the application. 
As it reduces the update rate, atomic output is not enabled by default. 
.SH OPTIONS
.TP
\fB\-n\fR \fInum\fR, \fB\-\-number\fR=\fInum\fR
Display only the top \fInum\fR entries of the histogram (default: all entries).
.TP
\fB\-i\fR \fIseconds\fR, \fB\-\-interval\fR=\fIseconds\fR
The refresh interval in seconds. The default is 0, updating the output as fast as possible.
.TP
\fB\-a\fR, \fB\-\-atomic\fR
Ensure that even on Ctrl+C, the application finishes printing the histogram before exiting.
.TP
\fB\-o\fR \fIname\fR, \fB\-\-out\fR=\fIname\fR
Output latest histogram to file \fIfname\fR.
.TP
\fB\-p\fR, \fB\-\-percent\fR
Show symbol frequency as percentage instead of absolute number.
.TP
\fB\-s\fR \fIsymbol\fR, \fB\-\-seperator\fR=\fIsymbol\fR
Set the separator which is displayed between the symbol and count. The default symbol is "\fB :: \fR". 
.TP
\fB\-c\fR, \fB\-\-no\-clear\fR 
Disable clearing of screen.
.TP
\fB\-r\fR, \fB\-\-reverse\fR
Sort ascending instead of descending.
.TP
\fB\-x\fR, \fB\-\-switch\fR
Show count first, then symbol.
.TP
\fB\-d\fR, \fB\-\-draw\fR
Draw bar for visualization.
.TP
\fB\-v\fR, \fB\-\-version\fR
Show version and exit.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show the help and exit.
.SH EXAMPLES
.TP 
cat file | livehist
Displays a histogram where every line of the file is seen as one element
.TP
python3 -c 'import random; [print(random.choice("ACABACAAABAABC")) for x in range(10000)]' | livehist
Randomly prints the letters A, B, and C using python and feeds it into \fBlivehist\fR. A sample output would be

    5763 :: A |###############################################################
    2137 :: C |#######################
    2100 :: B |#######################


.SH COPYRIGHT
Copyright 2020 Michael Schwarz <michael.schwarz91@gmail.com>

This is free software; see the  source  for  copying  conditions.
There  is  NO  warranty;  not  even  for
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 


livehist.cpp
#include <iostream>
#include <string>
#include <map>
#include <set>
#include <functional>
#include <algorithm>
#include <cctype>
#include <locale>
#include <vector>
#include <chrono>
#include <memory.h>
#include <unistd.h>
#include <getopt.h>
#include <signal.h>
#include "cli_parser.h"

// ---------------------------------------------------------------------------
static getopt_arg_t cli_options[] =
        {
                {"separator", required_argument, NULL, 's', "Set the seperator between symbol and count (default: \" :: \")", "SEPERATOR"},
                {"number",    required_argument, NULL, 'n', "Display only top NUMBER entries (default: all)", "NUMBER"},
                {"no-clear",  no_argument,       NULL, 'c', "Disable screen clear.", NULL},
                {"reverse",   no_argument,       NULL, 'r', "Short ascending.",      NULL},
                {"percent",   no_argument,       NULL, 'p', "Show frequency as percentage.", NULL},
                {"draw",      no_argument,       NULL, 'd', "Draw bar for visualization.", NULL},
                {"interval",  required_argument, NULL, 'i', "Refresh interval in seconds (default: 0).", "INTERVAL"},
                {"switch",    no_argument,       NULL, 'x', "Show count first, then value.", NULL},
                {"outfile",   required_argument, NULL, 'o', "Output latest histogram to file FILE.", "FILE"},
                {"atomic",    no_argument,       NULL, 'a', "On Ctrl+C, application finishes printing the histogram.", NULL},
                {"version",   no_argument,       NULL, 'v', "Show version and exit.", NULL},
                {"help",      no_argument,       NULL, 'h', "Show this help.",       NULL},
                {NULL, 0,                        NULL, 0, NULL,                      NULL}
        };


// ---------------------------------------------------------------------------
static volatile int terminated = 0;

// ---------------------------------------------------------------------------
static void interruption_handler(int signum) {
  if(signum == SIGINT) {
    terminated = 1;
  }
}

// ---------------------------------------------------------------------------
static void atomic_start() {
    terminated = 0;
    signal(SIGINT, interruption_handler);
}

// ---------------------------------------------------------------------------
static void atomic_end() {
    signal(SIGINT, SIG_DFL);
    fflush(stdout);
    if(terminated) {
        exit(130);
    }
}


// ---------------------------------------------------------------------------
static inline void trim(std::string &s) {
    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](int ch) {
        return !std::isspace(ch);
    }));
    s.erase(std::find_if(s.rbegin(), s.rend(), [](int ch) {
        return !std::isspace(ch);
    }).base(), s.end());
}

// ---------------------------------------------------------------------------
static bool sorter(std::pair<std::string, size_t>& elem1, std::pair<std::string, size_t>& elem2) {
    return elem1.second > elem2.second;
}


// ---------------------------------------------------------------------------
static bool rev_sorter(std::pair<std::string, size_t>& elem1, std::pair<std::string, size_t>& elem2) {
    return elem1.second < elem2.second;
}

// ---------------------------------------------------------------------------
static size_t update(Options& options, std::string& line, std::map<std::string, size_t>& hist, std::vector<std::pair<std::string, size_t>>& pairs) {
    // update histogram
    trim(line);
    if(hist.find(line) != hist.end()) {
        hist[line]++;
    } else {
        hist[line] = 1;
    }

    size_t sum = 0;
    pairs.clear();
    for (auto itr = hist.begin(); itr != hist.end(); ++itr) {
        pairs.push_back(*itr);
        sum += itr->second;
    }

    // sort bins
    sort(pairs.begin(), pairs.end(), options.reverse ? rev_sorter : sorter);
    return sum;
}

// ---------------------------------------------------------------------------
static void print_line(FILE* f, char* buffer, size_t maxlen, int draw, size_t val, size_t sum) {
    size_t len = fprintf(f, "%s", buffer);
    if(draw) {
        while(len++ < maxlen) {
            fprintf(f, " ");
        }
        fprintf(f, " |");
        for(unsigned int i = 0; i < ((120 - maxlen) * val) / sum; i++) fprintf(f, "#");
    }
    fprintf(f, "\n");
}

// ---------------------------------------------------------------------------
static void clear_screen(int vt100) {
    if(vt100) {
        std::cout << "\033[2J";
    } else {
        std::cout << std::endl;
    }
}

// ---------------------------------------------------------------------------
static void draw(Options& options, int output, FILE* out, std::vector<std::pair<std::string, size_t>>& pairs, size_t sum) {
    char buffer[257], tmp[256];
    size_t maxlen = 0;
    size_t limit = options.number;
    for (std::pair<std::string, size_t> element : pairs) {
        if(options.number && !limit) {
            break;
        }
        limit--;
        size_t len = 0;
        if(options.order) {
            len += snprintf(buffer, sizeof(buffer), "%s%s", element.first.c_str(), options.seperator.c_str());
            if(options.percent) len += snprintf(tmp, sizeof(tmp), "%f", element.second * 100.0 / sum);
            else len += snprintf(tmp, sizeof(tmp), "%zd", element.second);
            strncat(buffer, tmp, sizeof(buffer) - 1);
        } else {
            if(options.percent) len += snprintf(buffer, sizeof(buffer), "%f", element.second * 100.0 / sum);
            else len += snprintf(buffer, sizeof(buffer), "%zd", element.second);
            len += snprintf(tmp, sizeof(tmp), "%s%s", options.seperator.c_str(), element.first.c_str());
            strncat(buffer, tmp, sizeof(buffer) - 1);
        }
        if(len > maxlen) maxlen = len;
        if(output) {
            print_line(stdout, buffer, maxlen, options.draw, element.second, sum);

            if(options.outfile && out) {
                print_line(out, buffer, maxlen, options.draw, element.second, sum);
            }

        }
    }
}

// ---------------------------------------------------------------------------
static void draw_histogram(Options& options, std::vector<std::pair<std::string, size_t>>& pairs, size_t sum) {
    FILE* out = NULL;

    for(int output = 0; output < 2; output++) {
        if(output && (options.outfile || options.atomic))  {
            atomic_start();
            if(options.outfile) out = fopen(options.outfile, "w");
        }

        do {
            draw(options, output, out, pairs, sum);

            if(output && (options.outfile || options.atomic)) {
                if(options.outfile) {
                    fclose(out);
                }
                if(options.atomic) {
                    // on atomic update, redraw the histogram and exit afterwards
                    if(terminated == 1) {
                        terminated = 2;
                        clear_screen(options.clear);
                        continue;
                    }
                }
                atomic_end();
            }
            break;
        } while(options.atomic);
    }
}

// ---------------------------------------------------------------------------
int main(int argc, char* argv[]) {
    Options options;
    size_t last_update = 0;
    std::map<std::string, size_t> hist;
    std::vector<std::pair<std::string, size_t>> pairs;

    if(parse_cli_options(options, cli_options, argc, argv)) {
        return 1;
    }
    
    if(options.version) {
        printf("Live Histogram %s\n\nWritten by Michael Schwarz\n", APPVERSION);
        return 0;
    }

    for (std::string line; getline(std::cin, line);) {
        size_t sum = update(options, line, hist, pairs);
        if(!sum) sum = 1;

        // output
        size_t now = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();

        if(!options.interval || now - last_update >= options.interval * 1000) {
            last_update = now;
            clear_screen(options.clear);
            draw_histogram(options, pairs, sum);
        }
    }
}
